src/Analyzer/CodeAnalyzer.cs

```diff
@@ -0,0 +1,171 @@
+            }
+            else
+            {
+                tsqllintConfig.Rules = new Dictionary<string, string>();
+                foreach (var pattern in Patterns.Patterns.Select(p => p.PatternId))
+                {
+                    tsqllintConfig.Rules.Add(pattern, "error");
+                }
+            }
+
+            if (tsqllintConfig.CompatibilityLevel is null)
+            {
+                tsqllintConfig.CompatibilityLevel = 120;
+            }
+
+            File.WriteAllText(tmpTSQLLintPath, tsqllintConfig.ToString());
+        }
+
+        /// <summary>
+        /// Free temporary resources created on object construction.
+        /// </summary>
+        ~CodeAnalyzer()
+        {
+            // delete created temporary directory
+            Directory.Delete(Path.GetDirectoryName(tmpTSQLLintPath), true);
+        }
+
+        /// <summary>
+        /// Run analyze task
+        /// </summary>
+        /// <param name="cancellationToken">task cancellation token</param>
+        /// <returns>Task of the tool running</returns>
+        protected override async Task Analyze(CancellationToken cancellationToken)
+        {
+            await Task.Run(RunTool, cancellationToken);
+        }
+
+
+        /// <summary>
+        /// Run the actual tool.
+        /// This will call the rule processors of the tool, handle the available plugins
+        /// and analyze every file.
+        /// </summary>
+        private void RunTool()
+        {
+            timer.Start();
+
+            configReader.LoadConfig(tmpTSQLLintPath);
+
+            // tool write to Environment.ExitCode when occur an error
+            if (Environment.ExitCode != 0)
+            {
+                Console.WriteLine("Unexpected error. Check your config file!");
+                Environment.Exit(1);
+            }
+
+            var fragmentBuilder = new FragmentBuilder(configReader.CompatabilityLevel);
+            var ruleVisitorBuilder = new RuleVisitorBuilder(configReader, this.reporter);
+            var ruleVisitor = new SqlRuleVisitor(ruleVisitorBuilder, fragmentBuilder, reporter);
+            pluginHandler = new PluginHandler(reporter);
+            fileProcessor = new SqlFileProcessor(ruleVisitor, pluginHandler, reporter, new FileSystem());
+
+            pluginHandler.ProcessPaths(configReader.GetPlugins());
+
+            // tool write to Environment.ExitCode when occur an error
+            if (Environment.ExitCode != 0)
+            {
+                Console.WriteLine("Unexpected error. Probably some plugin not loaded properly!");
+                Environment.Exit(1);
+            }
+
+            Parallel.ForEach(Config.Files, (file) =>
+            {
+                try
+                {
+                    fileProcessor.ProcessPath(DefaultSourceFolder + file);
+                } catch (Exception e)
+                {
+                    Logger.Send(e);
+
+                    Console.WriteLine(new CodacyResult

```

Comment:
I don't see a good benefit on this, because the tool use provides this custom console reporter, which i'm using to report the codacy format.

Check this https://github.com/codacy/codacy-tsqllint/blob/2c92fa0b4e441d8c507150373ac4131ef406559b/src/Analyzer/Reporters/CodacyConsoleReporter.cs .

Reply:
this has a couple of drawbacks:
 - it's extremely hard to test (I don't think it's tested at all)
 - running within a `Parallel.ForEach` block will produce possibly inconsistent results from time to time and we don't have instruments to fix this

please aggregate the result and print it at the end.

Reply:
Done